dev/AppNotifications/AppNotificationManager.cpp

```diff
@@ -212,145 +258,297 @@ namespace winrt::Microsoft::Windows::AppNotifications::implementation
+        try
+        {
+            winrt::com_ptr<::ABI::Microsoft::Internal::ToastNotifications::INotificationProperties> notificationProperties = winrt::make_self<NotificationProperties>(notification);
+
+            winrt::com_ptr<::ABI::Microsoft::Internal::ToastNotifications::INotificationTransientProperties> notificationTransientProperties = winrt::make_self<NotificationTransientProperties>(notification);
 
-        auto notificationAppId{ ::Microsoft::Windows::AppNotifications::Helpers::RetrieveNotificationAppId() };
+            auto notificationAppId{ ::Microsoft::Windows::AppNotifications::Helpers::RetrieveNotificationAppId() };
 
-        DWORD notificationId = 0;
-        THROW_IF_FAILED(ToastNotifications_PostToast(notificationAppId.c_str(), notificationProperties.get(), notificationTransientProperties.get(), &notificationId));
+            DWORD notificationId = 0;
+            THROW_IF_FAILED(ToastNotifications_PostToast(notificationAppId.c_str(), notificationProperties.get(), notificationTransientProperties.get(), &notificationId));
 
-        THROW_HR_IF(E_UNEXPECTED, notificationId == 0);
+            THROW_HR_IF(E_UNEXPECTED, notificationId == 0);
 
-        winrt::Microsoft::Windows::AppNotifications::implementation::AppNotification* notificationImpl = get_self< winrt::Microsoft::Windows::AppNotifications::implementation::AppNotification>(notification);
-        notificationImpl->SetNotificationId(notificationId);
+            winrt::Microsoft::Windows::AppNotifications::implementation::AppNotification* notificationImpl = get_self< winrt::Microsoft::Windows::AppNotifications::implementation::AppNotification>(notification);
+            notificationImpl->SetNotificationId(notificationId);
+        }
+        catch (...)
+        {
+            hr = wil::ResultFromCaughtException();
+        }
     }
 
     winrt::Windows::Foundation::IAsyncOperation<winrt::Microsoft::Windows::AppNotifications::AppNotificationProgressResult> AppNotificationManager::UpdateAsync(winrt::Microsoft::Windows::AppNotifications::AppNotificationProgressData const data, hstring const tag, hstring const group)
     {
         THROW_HR_IF_MSG(E_INVALIDARG, tag == winrt::hstring(L""), "Update operation isn't guaranteed to find a specific notification to replace correctly.");
         THROW_HR_IF_MSG(E_INVALIDARG, data.SequenceNumber() == 0, "Sequence Number for Updates should be greater than 0!");
 
+        HRESULT hr{ S_OK };
+        std::wstring appId{};
+
         auto strong = get_strong();
         co_await resume_background();
 
-        std::wstring appId{ RetrieveNotificationAppId() };
+        auto logTelemetry{ wil::scope_exit([&]() {
+            AppNotificationTelemetry::UpdateAsyncByAPI(hr, appId);
+            THROW_HR_IF(hr, !(hr == S_OK || hr == E_NOT_SET));
```

Comment:
This is an example of why I'm a proponent of keeping success and failed exit points separate.

Reply:
Also resolved.